crates/sui-core/src/authority_server.rs

```diff
@@ -529,15 +535,100 @@ impl ValidatorService {
             .state
             .execute_certificate(&certificate, &epoch_store)
             .await?;
-        let events = if let Some(event_digest) = effects.events_digest() {
-            self.state.get_transaction_events(event_digest)?
+        let events = if request.include_events {
+            if let Some(digest) = effects.events_digest() {
+                Some(self.state.get_transaction_events(digest)?)
+            } else {
+                None
+            }
         } else {
-            TransactionEvents::default()
+            None
         };
-        Ok(Some(HandleCertificateResponseV2 {
-            signed_effects: effects.into_inner(),
+
+        let input_objects = request
+            .include_input_objects
+            .then(|| {
+                // Note unwrapped_then_deleted contains **updated** versions.
+                let unwrapped_then_deleted_obj_ids = effects
+                    .unwrapped_then_deleted()
+                    .into_iter()
+                    .map(|k| k.0)
+                    .collect::<HashSet<_>>();
+
+                let input_object_keys = effects
+                    .input_shared_objects()
+                    .into_iter()
+                    .map(|kind| {
+                        let (id, version) = kind.id_and_version();
+                        ObjectKey(id, version)
+                    })
+                    .chain(
+                        effects
+                            .modified_at_versions()
+                            .into_iter()
+                            .map(|(object_id, version)| ObjectKey(object_id, version)),
+                    )
+                    .collect::<HashSet<_>>()
+                    .into_iter()
+                    // Unwrapped-then-deleted objects are not stored in state before the tx, so we have nothing to fetch.
+                    .filter(|key| !unwrapped_then_deleted_obj_ids.contains(&key.0))
+                    .collect::<Vec<_>>();
```

Comment:
I think just calling `modified_at_versions` should already be sufficient. It will include any modified shared objects, and it will not include unwrapped_then_deleted_obj_ids.

Reply:
Has that been changed? we needed to do this elsewhere (indexer) when getting input/output objects because we had unwrapped_then_deleted objects not existing in the DB causing reads to fail.

Reply:
Prior to `simplified_unwrapped_then_deleted`, it was possible for `modified_at_versions` to contain `unwrapped_then_deleted` objects, which would never exist in the DB. That's long gone now.